### PR TITLE
[Fix-3616][Server] when worker akc/response master exception , async retry

### PR DESCRIPTION
**Pull request by qiaozhanwei**
fix bug #3616 on 1.3.3-release

### CODE REVIEW

**Review by lgcareer — APPROVED** [review]
+1

File: dolphinscheduler-remote/src/main/java/org/apache/dolphinscheduler/remote/command/CommandType.java [hunk 1]
```diff
@@ -1 +1 @@
...
    /**
     *  ping
     */
    PING,

    /**
     *  pong
     */
    PONG;
}
```
Comment (dailidong) [inline dolphinscheduler-remote/src/main/java/org/apache/dolphinscheduler/remote/command/CommandType.java:1]: pelase remove ^M

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/runner/RetryReportTaskStatusThread.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    /**
+     * every 5 minutes
+     */
+    private static long RETRY_REPORT_TASK_STATUS_TIME = 5 * 60 * 1000L;
+    /**
+     *  task callback service
+     */
+    private final TaskCallbackService taskCallbackService;
+
+    public void start(){
```
Comment (Technoboy-) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/runner/RetryReportTaskStatusThread.java:50]: Which service starts this ?

**Comment by Technoboy-** [comment]
I think this is just a fix for #3616 .  It seems good.
For #3616 , it looks a simple issue, but involves that Master must have the ability to ping the worker as the task times-out.

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/runner/RetryReportTaskStatusThread.java (hunk 2 above)
Comment (qiaozhanwei) [reply dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/runner/RetryReportTaskStatusThread.java:50]: // retry report task status
        this.retryReportTaskStatusThread.start();

File: dolphinscheduler-remote/src/main/java/org/apache/dolphinscheduler/remote/command/CommandType.java (hunk 1 above)
Comment (qiaozhanwei) [reply dolphinscheduler-remote/src/main/java/org/apache/dolphinscheduler/remote/command/CommandType.java]: resloved

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/cache/ResponceCache.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+package org.apache.dolphinscheduler.server.worker.cache;
+
+import org.apache.dolphinscheduler.common.enums.Event;
+import org.apache.dolphinscheduler.remote.command.Command;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class ResponceCache {
```
Comment (Technoboy-) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/cache/ResponceCache.java:26]: Miss java doc

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/processor/NettyRemoteChannel.java [hunk 4]
```diff
@@ -43,13 +45,21 @@
...
     public NettyRemoteChannel(Channel channel, long opaque) {
         this.channel = channel;
         this.host = ChannelUtils.toAddress(channel);
         this.opaque = opaque;
     }
 
+    public NettyRemoteChannel(Channel channel) {
+        this.channel = channel;
+        this.host = ChannelUtils.toAddress(channel);
+        this.opaque = random.nextLong();
```
Comment (Technoboy-) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/processor/NettyRemoteChannel.java:60]: opaque should be -1 is much better.  we can locate the issue easily.

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/cache/ResponceCache.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+    public static ResponceCache get(){
+        return instance;
+    }
+
+    private Map<Integer,Command> ackCache = new ConcurrentHashMap<>();
+    private Map<Integer,Command> responseCache = new ConcurrentHashMap<>();
+
+
+    public void cache(Integer taskInstanceId, Command command, Event event){
```
Comment (Technoboy-) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/cache/ResponceCache.java:40]: Rename to add

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/runner/RetryReportTaskStatusThread.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    public RetryReportTaskStatusThread(){
+        this.taskCallbackService = SpringApplicationContext.getBean(TaskCallbackService.class);
+    }
+    @Override
+    public void run() {
+        ResponceCache responceCache = ResponceCache.get();
+
+        while (Stopper.isRunning()){
+            if (responceCache.getAckCache().isEmpty() && responceCache.getResponseCache().isEmpty()){
+                continue;
```
Comment (Technoboy-) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/runner/RetryReportTaskStatusThread.java:64]: Should sleep.

Comment (qiaozhanwei) [reply dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/runner/RetryReportTaskStatusThread.java:64]: ok

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/cache/ResponceCache.java (hunk 3 above)
Comment (qiaozhanwei) [reply dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/cache/ResponceCache.java:26]: ok

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/processor/NettyRemoteChannel.java (hunk 4 above)
Comment (qiaozhanwei) [reply dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/processor/NettyRemoteChannel.java:60]: ok

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/cache/ResponceCache.java (hunk 5 above)
Comment (qiaozhanwei) [reply dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/cache/ResponceCache.java:40]: cache maybe express my idea

Comment (Technoboy-) [reply dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/cache/ResponceCache.java:49]: For the class Name XXXCache.

**Review by Technoboy- — APPROVED** [review]

**Pull request merged**